[WPE][GTK] Can't paste image data (like image/png) from the clipboard in WebKit browsers

mcatanzaro opened the pull request:
#### d7d1107bc9b63b1a7aef053d6f900cc81bd28204
<pre>
[WPE][GTK] Can&apos;t paste image data (like image/png) from the clipboard in WebKit browsers
<a href="https://bugs.webkit.org/show_bug.cgi?id=218519">https://bugs.webkit.org/show_bug.cgi?id=218519</a>

Reviewed by Adrian Perez de Castro.

Change the Pasteboard::read PasteboardFileReader overload to more
closely match the Cocoa implementation. Notably, we now accept more
image types as input rather than supporting only PNG. However, our
implementation is simpler because we only support a single clipboard
item, so the index parameter is ignored.

Also, fix a bug that caused us to bail early if
PasteboardStrategy::readFilePathsFromClipboard does not return any
filenames.

Thanks to Miikka, Sonny Piers, and 荒野無燈 for their help in the bug
report.

Also, I&apos;ve snuck in a couple extra fixes in other files, to make the
Pasteboard constructor explicit and fix a typo.

Canonical link: <a href="https://commits.webkit.org/301877@main">https://commits.webkit.org/301877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/32e98b2eec7b2316f611bfe3e995d6e7024f296a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126644 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46288 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37250 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133613 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78303 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128515 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46921 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54826 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96378 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64467 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37539 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113274 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76904 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36427 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31457 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77009 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107354 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31752 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136183 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53331 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41032 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104890 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53820 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109629 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104589 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26800 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50075 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28418 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50763 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53253 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59066 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52530 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55866 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54267 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/glib/PasteboardGLib.cpp:
```diff
@@ -334,14 +356,33 @@ void Pasteboard::read(PasteboardFileReader& reader, std::optional<size_t> index)
+            if (!reader.shouldReadBuffer(imageType))
+                continue;
+            auto buffer = readBuffer(itemIndex, imageType);
+            if (buffer) {
```
I should probably write this as:

```suggestion
            if (auto buffer = readBuffer(itemIndex, imageType)) {
```

mcatanzaro commented on Source/WebCore/platform/glib/PasteboardGLib.cpp:
```diff
@@ -334,14 +356,33 @@ void Pasteboard::read(PasteboardFileReader& reader, std::optional<size_t> index)
+                continue;
+            auto buffer = readBuffer(itemIndex, imageType);
+            if (buffer) {
+                reader.readBuffer(imageTypeToFakeFilename(imageType), imageType, buffer.releaseNonNull());
```
Than can probably be:

```suggestion
                reader.readBuffer(imageTypeToFakeFilename(imageType), imageType, WTFMove(buffer));
```

I'll investigate before landing.

mcatanzaro replied:
No, it was right. I was afraid that `releaseNonNull()` was leaking, but it actually just converts RefPtr to Ref.

carlosgcampos commented on Source/WebCore/platform/glib/PasteboardGLib.cpp:
```diff
@@ -334,14 +356,33 @@ void Pasteboard::read(PasteboardFileReader& reader, std::optional<size_t> index)
         auto filePaths = platformStrategies()->pasteboardStrategy()->readFilePathsFromClipboard(m_name);
-        for (const auto& filePath : filePaths)
-            reader.readFilename(filePath);
+        if (filePaths.size() > 0) {
```
```suggestion
        if (!filePaths.isEmpty()) {
```

carlosgcampos commented on Source/WebCore/platform/glib/PasteboardGLib.cpp:
```diff
@@ -334,14 +356,33 @@ void Pasteboard::read(PasteboardFileReader& reader, std::optional<size_t> index)
+    };
+
+    if (index) {
+        readBufferAtIndex(*index);
```
I guess this works because index is 0. We don't support more than one item, in the pasteboard, `WebPasteboardProxy::readBufferFromPasteboard` returns early if index != 0, so we can save IPC calls by checking too and only call readBufferAtIndex is index == 0

mcatanzaro replied:
Simplified.

carlosgcampos commented on Source/WebCore/platform/glib/PasteboardGLib.cpp:
```diff
@@ -334,14 +356,33 @@ void Pasteboard::read(PasteboardFileReader& reader, std::optional<size_t> index)
-            reader.readBuffer({ }, { }, buffer.releaseNonNull());
+    if (auto allInfo = allPasteboardItemInfo()) {
+        for (size_t itemIndex = 0; itemIndex < allInfo->size(); ++itemIndex)
+            readBufferAtIndex(itemIndex);
```
This is ok because allPasteboardItemInfo always returns a Vector of size = 1

mcatanzaro replied:
Removed.

aperezdc commented on Source/WebCore/platform/glib/PasteboardGLib.cpp:
```diff
@@ -266,6 +266,29 @@ void Pasteboard::setDragImage(DragImage, const IntPoint&)
+{
+    static std::array<ASCIILiteral, 6> types = { "image/png"_s, "image/jpeg"_s, "image/gif"_s, "image/bmp"_s, "image/vnd.microsoft.icon"_s, "image/x-icon"_s };
+    return types;
+}
```
This does not need to be a function, it can be a constant global:


```suggestion
static constexpr std::array<ASCIILiteral, 6> imageTypes = { "image/png"_s, "image/jpeg"_s, "image/gif"_s, "image/bmp"_s, "image/vnd.microsoft.icon"_s, "image/x-icon"_s };
```

mcatanzaro replied:
Done.

mcatanzaro committed:
[WPE][GTK] Can't paste image data (like image/png) from the clipboard in WebKit browsers
https://bugs.webkit.org/show_bug.cgi?id=218519

Reviewed by Adrian Perez de Castro.

Change the Pasteboard::read PasteboardFileReader overload to more
closely match the Cocoa implementation. Notably, we now accept more
image types as input rather than supporting only PNG. However, our
implementation is simpler because we only support a single clipboard
item, so the index parameter is ignored.

Also, fix a bug that caused us to bail early if
PasteboardStrategy::readFilePathsFromClipboard does not return any
filenames.

Thanks to Miikka, Sonny Piers, and 荒野無燈 for their help in the bug
report.

Also, I've snuck in a couple extra fixes in other files, to make the
Pasteboard constructor explicit and fix a typo.

Canonical link: https://commits.webkit.org/301877@main
